[Executorch] Add broadcasting support to optimized op_sub

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #8257
* __->__ #8256

Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc @larryliu0820 @manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

kimishpatel committed:
[Executorch] Refactor op_mul's broadcasting utils

Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
[ExecuTorch] Add broadcast support for optimized add op

Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
[Executorch] Refactor op_add to support op_sub broadcasting

Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
[Executorch] Add broadcasting support to optimized op_sub

Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

swolchok commented on kernels/test/op_sub_test.cpp:
```diff
@@ -99,6 +99,109 @@ class OpSubOutTest : public OperatorTest {
+    TensorFactory<DTYPE> tf_a;
+
+    Tensor a =
+        tf_a.make({2, 2, 3}, /*data=*/{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12});
```
as before, programmatically filling test data would be nice I think

kimishpatel replied:
yeah I agree. Didnt quite occur to me at the time of writing but thanks for pointing out

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69491818).

kimishpatel committed:
Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69491818).

kimishpatel committed:
Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69491818).

kimishpatel committed:
Update base for Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69491818).

kimishpatel committed:
Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69491818).

kimishpatel committed:
Update base for Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Add broadcasting support to optimized op_sub"


Summary:
This diff builds on top of previous one to add support for limited
handling of broadcasting for sub

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491818](https://our.internmc.facebook.com/intern/diff/D69491818)

[ghstack-poisoned]
